core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp

```diff
@@ -51,11 +51,21 @@ struct GraphImpl<Kokkos::Cuda> {
 
   using node_details_t = GraphNodeBackendSpecificDetails<Kokkos::Cuda>;
 
+  using erased_node_impl_t =
+      GraphNodeImpl<Kokkos::Cuda, Kokkos::Experimental::TypeErasedTag,
+                    Kokkos::Experimental::TypeErasedTag>;
+
   // Store drivers for the kernel nodes that launch in global memory.
   // This is required as lifetime of drivers must be bounded to this instance's
   // lifetime.
   std::vector<std::shared_ptr<void>> m_driver_storage;
 
+  // Store 'then' nodes. This is required because we don't enforce the user to
+  // keep a reference to 'then' nodes, so we need to guarantee that someone (us)
+  // stores the functor used for the capture, that might be keeping some objects
+  // alive.
+  std::vector<std::shared_ptr<erased_node_impl_t>> m_then_nodes;
```

Comment:
The default implementation session the `GraphNodeBackendSpecificDetails` as the common type for the storage. It could be an option here too, ie., to use the `node_details_t` type.

- https://github.com/kokkos/kokkos/blob/9f7d134bd57a3e1fced0e7b799042a26317522d4/core/src/impl/Kokkos_Default_Graph_Impl.hpp#L53

Reply:
Since `GraphNodeImpl` derives from `GraphNodeBackendSpecificDetails` https://github.com/kokkos/kokkos/blob/4889d4753daea6975d28589e662fa90825f6b88c/core/src/impl/Kokkos_GraphNodeImpl.hpp#L40-L44 we would probably store less if we keep `GraphNodeBackendSpecificDetails` (*e.g.* we would not store what `ExecutionSpaceInstanceStorage` stores).

Reply:
Done. Let's see if it works.